### ISSUE TITLE
Saints Row IV black screen with UI elements showing up

### ISSUE THREAD

**Issue by adih2001** [doitsujin/dxvk#1136]
### Software information
Name: Saints Row IV
Settings used: 
[https://caustic-rays-cdn.s3.eu-west-2.amazonaws.com/pics/dxvk/1.png](https://caustic-rays-cdn.s3.eu-west-2.amazonaws.com/pics/dxvk/1.png)
[https://caustic-rays-cdn.s3.eu-west-2.amazonaws.com/pics/dxvk/2.png](https://caustic-rays-cdn.s3.eu-west-2.amazonaws.com/pics/dxvk/2.png)
### System information
- GPU: GTX 970
- Driver: 431.36.0
- Wine version: Windows
- DXVK version: 1.2.5 (Working-ish) / 1.3 and up (not working)

### Apitrace file(s)
- [https://drive.google.com/open?id=19cA0M1ttkyscoDpDzT55DZFCGYTMZPQv](https://drive.google.com/open?id=19cA0M1ttkyscoDpDzT55DZFCGYTMZPQv)

### Log files
- d3d11.log: Its included in the link above ^^^
- dxgi.log: ^^^

### Notes:
I first tested this game with DXVK 1.2.3. The game had some flickering but the game worked overall as you can see here: [https://youtu.be/eB6HICBrcFM](https://youtu.be/eB6HICBrcFM) (Note: The effects on the walls are normal and they are from the game. They can be best seen at 1:08).
When I switched to DXVK 1.3 everything was completely black other then the UI elements as you can see here: [https://youtu.be/hj0xvrpiHIE](https://youtu.be/hj0xvrpiHIE). DXVK 1.3.1 does the exact same thing. 
In the google drive link I've included all the necessary files.

I know Windows is not a supported platform, so it would be great if someone could try and replicate the issue on Linux using the provided trace.

Thanks 🐸

**Comment by Joshua-Ashton** [comment]
Something really bizarre is going on here...

I **don't** get the issue when playing back the trace with native or with DXVK 1.3.1.
![image](https://user-images.githubusercontent.com/21316711/61586383-60c64380-ab6a-11e9-9f80-a6a0b64d3899.png)

I **do**, however, get the issue when playing back the trace with RenderDoc attached, and when trying to capture it crashes for me.
![image](https://user-images.githubusercontent.com/21316711/61586415-332dca00-ab6b-11e9-8d47-4dd65e2ef80f.png)

I spoke with @adih2001 and the trace plays back fine for him with DXVK 1.3.1 both with or without RenderDoc.

The validation layer output was not particularly interesting (to me anyway,) just the standard one we ignore for unbound resources.
```
10105268: debug: UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated(ERROR / SPEC): msgNum: 0 - Descriptor set 0x4cb88 bound as set #0 encountered the following validation error at vkCmdDraw() time: Descriptor in binding #1 index 0 requires FLOAT component type, but bound descriptor format is VK_FORMAT_R32_UINT.
    Objects: 1
       [0]  0x4cb88, type: 23, name: NULL
UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated(ERROR / SPEC): msgNum: 0 - Descriptor set 0x4cb88 bound as set #0 encountered the following validation error at vkCmdDraw() time: Descriptor in binding #1 index 0 requires FLOAT component type, but bound descriptor format is VK_FORMAT_R32_UINT.
    Objects: 1
        [0] 0x4cb88, type: 23, name: NULL
```

To me it looks like a lack of textures as sometimes I see dust and stuff which is correctly shown and depth culled. Which makes me think that perhaps there is something more to
```
err:   D3D11: CopySubresourceRegion: Incompatible texel size
  Dst texel size: 8
  Src texel size: 4
```
in the logs.

It's worth noting that the trace was made with native D3D11 on Windows.

I'm going to sleep now however, it's 4am! 🐸

**Comment by doitsujin** [comment]
Can anyone bisect this?

Note that the engine is known to be broken, so issues with DXVK are kind of expected at this point.

**Comment by tannisroot** [comment]
https://github.com/doitsujin/dxvk/commit/03c6df56
is what caused this issue
Note that I can only reproduce this with DXVK compiled using GCC. With MSVC build the issue does not occur.

**Comment by ryao** [comment]
@tannisroot I suspect that there is a bug in DXVK involving undefined behavior given your report on discord that an -O0 GCC build and that undefined behavior bugs being triggered by compiler optimizations on GCC are far more common than actual compiler bugs. MSVC's optimizer tends to be more friendly to undefined behavior bugs, so it producing builds that work doesn't contradict that theory. Here is a patch that should work around it:

https://github.com/ryao/dxvk/commit/ae458360a2cbb372a01990d76e234d1b366d213d

Unless `D3D11Initializer::InitDeviceLocalTexture()` is in a performance critical code path, it should workaround this issue with zero performance loss. Let me know if it makes things work.

**Comment by ryao** [comment]
As an additional note, if that patch works, it *might* be possible to bisect the function itself using those pragma commands to find the line with undefined behavior.

**Comment by ryao** [comment]
@adih2001 I suggest trying out the new Nvidia vulkan beta driver version 435.19.03:

https://developer.nvidia.com/vulkan-driver

It claims to have a corruption fix for this game.

**Comment by edmondo** [comment]
The black screen seems to be fixed after upgrading to DXVK 1.4 on my setup (AMD RX470 (Polaris)).

**Comment by doitsujin** [comment]
Some users noted that this might have been some weird issue with the DXVK builds rather than a code issue, but I'm not sure if that's true.

Anyway, closing.
